File: tts-worker/src/uploader.py
"""The WAV leaves the machine by streamed PUT to a signed Supabase URL.

Round 12's discipline, in Python: the render worker learned that fetch()
retains the request body even when "streaming" and OOM'd on 1.2 GB uploads;
its fix was http.request + createReadStream. The equivalent here is
requests.put(data=<file object>): requests reads the Content-Length from
fstat and streams the file in chunks — constant memory for a ~130 MB WAV.

The bytes go worker -> Supabase Storage directly. The signed URL is created
app-side at submission and rides the job payload; the Cloudflare Workers app
layer never sees a byte of audio.
"""

from __future__ import annotations

import os


class UploadFailure(Exception):
    """Worded for the project failure message; detail goes in parentheses."""


def upload_wav(path: str, signed_url: str, put=None) -> int:
    """PUTs the finished WAV. Returns the byte count sent.

    `put` is a test seam (requests.put-compatible). Every failure is a worded
    UploadFailure: an upload that fails must fail the job loudly — the
    alternative is a project whose transcript exists but whose audio does not,
    which is the silent-partial class this design excludes.
    """
    if put is None:
        import requests

        put = requests.put

    size = os.path.getsize(path)
    try:
        with open(path, "rb") as body:
            response = put(
                signed_url,
                data=body,
                headers={
                    "Content-Type": "audio/wav",
                    "Content-Length": str(size),
                },
                timeout=600,
            )
    except Exception as err:  # noqa: BLE001 — every transport error is terminal here
        raise UploadFailure(
            f"The narration could not be saved to storage. Please try again. "
            f"(internal: upload transport error: {err})"
        ) from err

    if not (200 <= response.status_code < 300):
        body_excerpt = ""
        try:
            body_excerpt = response.text[:200]
        except Exception:  # noqa: BLE001
            pass
        raise UploadFailure(
            f"The narration could not be saved to storage. Please try again. "
            f"(internal: upload returned HTTP {response.status_code}: {body_excerpt})"
        )
    return size

File: motion-worker/src/uploader.js
/**
 * The finished MP4 leaves the box by streamed PUT to a signed URL created
 * app-side — the same Round-12 discipline as the render and tts workers:
 * https.request + createReadStream, Content-Length from stat, constant
 * memory. The Cloudflare app layer never carries a video byte.
 */

import fs from 'node:fs';
import https from 'node:https';
import http from 'node:http';
import { URL } from 'node:url';

export function uploadFile(filePath, signedUrl, { contentType = 'video/mp4' } = {}) {
  return new Promise((resolve, reject) => {
    const size = fs.statSync(filePath).size;
    const url = new URL(signedUrl);
    const transport = url.protocol === 'http:' ? http : https;
    const request = transport.request(
      url,
      {
        method: 'PUT',
        headers: { 'Content-Type': contentType, 'Content-Length': size },
        timeout: 600_000,
      },
      (response) => {
        const chunks = [];
        response.on('data', (c) => chunks.length < 8 && chunks.push(c));
        response.on('end', () => {
          if (response.statusCode >= 200 && response.statusCode < 300) {
            resolve(size);
          } else {
            reject(
              new Error(
                `The finished explainer could not be saved to storage (HTTP ${response.statusCode}). ` +
                  `Please try again. (internal: ${Buffer.concat(chunks).toString('utf8').slice(0, 200)})`,
              ),
            );
          }
        });
      },
    );
    request.on('timeout', () => request.destroy(new Error('upload timed out')));
    request.on('error', (err) =>
      reject(
        new Error(
          `The finished explainer could not be saved to storage. Please try again. (internal: ${err.message})`,
        ),
      ),
    );
    fs.createReadStream(filePath).pipe(request);
  });
}
